File: eval/revision_modos/revision_humana.py
"""Interfaz de revisión humana de respuestas guardadas por modo (#41)."""
import argparse
import html
import hashlib
import json
import re
from pathlib import Path

RECURSOS = Path(__file__).parent

def mejorar(page, base, ejemplos=()):
    base = Path(base)
    if 'id="human-app"' in page:
        raise ValueError('La interfaz ya está incorporada')
    manifest = json.loads((base / 'entrega/manifest.json').read_text())
    photos = [{k: p[k] for k in ('foto', 'archivo', 'sha256')} for p in manifest['fotos']]
    dataset = hashlib.sha256(json.dumps(photos, sort_keys=True).encode()).hexdigest()
    ids = {p['foto'] for p in photos}
    if not photos or any(e['foto'] not in ids for e in ejemplos):
        raise ValueError('Conjunto vacío o ejemplo de una foto ajena al conjunto')
    modes = {}
    for p in photos:
        modes[p['foto']] = {}
        for mode in ('bajo', 'medio', 'alto'):
            r = json.loads((base / 'analisis' / f"{p['foto']}-{mode}.json").read_text())['resultado']
            keys = lambda values: sorted({v.get('key') or v.get('codigo') for v in values if v.get('key') or v.get('codigo')})
            modes[p['foto']][mode] = {'confirmados': keys(r.get('problemas', []) + r.get('elementos_detectados', [])), 'posibles': keys(r.get('posibles', [])), 'estado': r.get('analisis_estado'), 'costo': r.get('costo_api'), 'consumo_completo': r.get('tokens_api_completos')}
    data = json.dumps({'conjunto': dataset, 'fotos': photos, 'modos': modes, 'ejemplos': [e['foto'] for e in ejemplos]}, ensure_ascii=False).replace('<', '\\u003c')
    css = (RECURSOS / 'revision-humana.css').read_text()
    js = (RECURSOS / 'revision-humana.js').read_text()
    header_end = page.index('</style>') + len('</style>')
    prefix, legacy = page[:header_end], page[header_end:]
    prefix = re.sub(r'(<head\b[^>]*>)', r'\1<meta charset="utf-8">', prefix, count=1, flags=re.I)
    prefix = re.sub(r'<title>.*?</title>', '<title>Ojo Urbano: revisión humana de fotos</title>', prefix)
    legacy = legacy.replace('</html>', '')
    legacy = legacy.replace("document.querySelectorAll('select,input')", "document.querySelectorAll('#legacy select,#legacy input')")
    if 'id="advertencia-referencia"' not in legacy:
        legacy = legacy.replace('<header>', '<header><p id="advertencia-referencia" class="human-warn">La referencia provisional tiene inconsistencias de contenedores y de objetos voluminosos. Estos porcentajes históricos no permiten juzgar precisión hasta revisar la referencia. Las etiquetas originales se conservan.</p>', 1)
    app = '''<section id="human-app" aria-label="Revisión humana"><div class="human-top"><div><div class="human-kicker">OJO URBANO / REVISIÓN HUMANA</div><h1>¿Qué ves en esta foto?</h1><p>Marcá tu lectura antes de revelar los resultados. Podés elegir varias categorías y dejar dudas.</p></div><div><span class="human-count" id="avance">0 / 100</span><p class="human-muted">fotos con revisión terminada</p></div></div>
<progress class="human-progress" id="progreso" max="100" value="0"></progress>
<div class="human-toolbar"><button id="exportar">Exportar mi revisión</button><button id="importar">Importar revisión</button><input id="archivo-revision" type="file" accept=".json,application/json" hidden><button id="ver-puntajes">Ver mis puntajes</button><button id="ver-informe">Ver informe anterior</button></div>
<p class="human-status" id="guardado" role="status" aria-live="polite"></p><p class="human-muted">Se guarda en este navegador, sin enviar tus decisiones al servidor. Exportá una copia al terminar cada tanda. Para cambiar de dispositivo o de enlace, importá ese archivo. Las fotos omitidas quedan pendientes.</p>
<div id="import-preview" class="human-warn" hidden></div>
<details id="guia"><summary>Criterios y ejemplos de referencia</summary><div class="human-card"><p><strong>Primero, lo visible.</strong> Escombros: restos de mampostería o construcción. Una bolsa cerrada sin señales suficientes queda como duda. Residuos comunes: bolsas, envases, papeles o reciclables. Poda: ramas cortadas y restos de jardinería, no vegetación viva.</p><p><strong>Voluminosos incluye más que muebles:</strong> aberturas, sanitarios, placas grandes retiradas y envases de pintura o químicos descartados. Cascotes o fragmentos de mampostería corresponden a escombros. Un objeto en uso no prueba que haya que retirarlo. Cartón no equivale a madera.</p><p><strong>Después, el servicio.</strong> Marcá Sí cuando la escena respalde el reclamo. Usá No evaluable si se ve el material pero no alcanza para decidir el servicio, por ejemplo en un interior. No conviertas una duda en No.</p><p><strong>Contenedores, según los criterios del proyecto:</strong> lateral de húmedos con cuerpo negro u oliva y paredes curvas; bilateral con cuerpo gris y paredes más planas. Los postes no distinguen por sí solos. Sombras y reflejos pueden engañar; ante duda, no fuerces el subtipo. Secos se refiere al contenedor identificado para reciclables. Puede haber varios tipos en la misma foto.</p><!-- EJEMPLOS --><p class="human-muted">Si estos criterios no coinciden con tu interpretación del servicio, explicalo en la nota. Los puntajes no incluyen cantidad de objetos, gravedad ni categorías fuera de esta lista.</p></div></details>
<div class="human-toolbar"><button id="anterior">Anterior</button><label>Foto <select id="foto-select"></select></label><button id="siguiente">Siguiente</button><button id="pendiente">Próxima pendiente</button></div>
<div class="human-grid"><section class="human-image"><a id="foto-grande" target="_blank" rel="noopener"><img id="foto-humana" alt=""><span>Abrir foto en tamaño completo</span></a></section><section class="human-card"><h2 id="foto-titulo"></h2><p id="foto-estado" class="human-muted"></p><form id="form-revision"><h3>1. Material u objeto visible</h3><p class="human-muted">Aunque esté en un interior o en uso.</p><div id="materiales"></div><h3>2. ¿Corresponde confirmar el servicio?</h3><p class="human-muted">Evaluá la escena, no la categoría del vecino.</p><div id="servicios"></div><h3>3. Contenedores presentes</h3><div id="contenedores"></div><label>Notas u otros problemas visibles<textarea id="notas" class="human-text" maxlength="3000" placeholder="Qué se ve, qué no se puede determinar o por qué cambiaste una decisión"></textarea></label><p><label><input type="checkbox" id="exposicion-previa"> Ya había visto resultados o etiquetas de esta foto antes de revisarla.</label></p><p id="exposicion-aviso" class="human-muted"></p><div class="human-toolbar"><button type="submit" class="human-primary" id="terminar">Terminar revisión y seguir</button><button type="button" id="revelar">Revelar resultados de esta foto</button></div><p id="form-mensaje" role="status" aria-live="polite"></p></form></section></div>
<section id="resultados-revelados" class="human-revealed" hidden><h2>Resultados guardados para esta foto</h2><p class="human-warn">La lectura provisional puede contener errores. Tu revisión se guarda aparte. Revelar las respuestas queda registrado.</p><div id="resultados-foto"></div></section>
<section id="puntajes" class="human-card human-metrics" hidden><h2>Puntajes con tu revisión</h2><p>Se calculan solo sobre fotos con revisión terminada. Precisión = confirmaciones correctas / todas las confirmaciones evaluables. Detección = positivos confirmados / positivos humanos evaluables. Un positivo dejado como posible cuenta como pendiente y reduce detección; no se suma a las omisiones. Las dudas y los servicios no evaluables quedan fuera de ambos denominadores.</p><p class="human-muted">El material visible se conserva para revisar reconocimiento, pero no se puntúa automáticamente contra descripciones de texto libre. Esta evaluación mide cuatro servicios y tres presencias de contenedor. No demuestra el aporte aislado del tercer modelo, porque Completo también incluye el especialista.</p><div id="puntajes-contenido"></div></section></section>'''
    imagenes = {p['foto']: p['archivo'] for p in photos}
    guia = '<div class="human-help-grid">' + ''.join(
        '<figure><img src="' + html.escape(imagenes[e['foto']], quote=True) +
        '" loading="lazy" alt="' + html.escape(e['descripcion'], quote=True) +
        '"><figcaption>' + html.escape(e['descripcion']) +
        '. Esta guía registra la exposición de esta foto.</figcaption></figure>'
        for e in ejemplos) + '</div>'
    app = app.replace('<!-- EJEMPLOS -->', guia)
    app = app.replace('0 / 100', f'0 / {len(photos)}').replace('max="100"', f'max="{len(photos)}"')
    return prefix + f'<style>{css}</style><body>' + app + '<div id="legacy" hidden><div class="human-legacy-exit"><button id="volver-revision">Volver a mi revisión</button></div>' + legacy + '</div><noscript>Activá JavaScript para completar y guardar tu revisión.</noscript><script type="application/json" id="human-data">' + data + '</script><script>' + js + '</script></body></html>'

if __name__ == '__main__':
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--base', required=True)
    parser.add_argument('--entrada', required=True, help='Informe anterior sin interfaz de revisión')
    parser.add_argument('--salida', required=True, help='HTML nuevo; no sobrescribe un archivo existente')
    parser.add_argument('--ejemplos', help='JSON privado opcional: lista de foto y descripcion')
    args = parser.parse_args()
    ejemplos = json.loads(Path(args.ejemplos).read_text()) if args.ejemplos else []
    page = mejorar(Path(args.entrada).read_text(), args.base, ejemplos)
    with Path(args.salida).open('x') as f:
        f.write(page)
    print('Interfaz de revisión humana creada, sin inferencias.')

File: eval/revision_modos/revision-humana.js
(()=>{
'use strict';
const D=JSON.parse(document.getElementById('human-data').textContent), $=id=>document.getElementById(id);
const services=['recoleccion','retiro_escombros','retiro_muebles','retiro_poda'], containers=['contenedor_humedos_lateral','contenedor_humedos_bilateral','contenedor_secos'];
const labels={recoleccion:'Residuos comunes / reciclables',retiro_escombros:'Escombros / restos de obra',retiro_muebles:'Muebles y otros voluminosos',retiro_poda:'Poda / jardinería',contenedor_humedos_lateral:'Húmedos lateral',contenedor_humedos_bilateral:'Húmedos bilateral',contenedor_secos:'Secos / reciclables'};
const modes={bajo:'Económico',medio:'Equilibrado',alto:'Completo'}, ids=D.fotos.map(p=>p.foto), KEY='ojo-revision-humana-v1:'+D.conjunto;
const blank=()=>({materiales:Object.fromEntries(services.map(k=>[k,''])),servicios:Object.fromEntries(services.map(k=>[k,''])),contenedores:Object.fromEntries(containers.map(k=>[k,''])),notas:'',completa:false,exposicion_previa:false,revelada_en:null,guia_en:null,terminada_en:null,actualizada_en:null,historial:[]});
let state={version:1,conjunto:D.conjunto,actual:ids[0],revisiones:{}}, shown=false, pendingImport=null, storageBroken=false, externalConflict=false;
const allowed={materiales:['','si','no','duda'],servicios:['','si','no','duda','no_evaluable'],contenedores:['','si','no','duda']};
const fields={materiales:services,servicios:services,contenedores:containers};
const get=id=>state.revisiones[id]||blank(), current=()=>get(state.actual), stamp=()=>new Date().toISOString();
const message=(text,error=false)=>{$('guardado').textContent=text;$('guardado').classList.toggle('error',error);};
const full=r=>Object.keys(fields).every(g=>fields[g].every(k=>r[g][k]!==''));
function validate(v){
 if(!v||v.version!==1||v.conjunto!==D.conjunto||!ids.includes(v.actual)||!v.revisiones||Array.isArray(v.revisiones)||typeof v.revisiones!=='object')throw Error('El archivo no corresponde a este conjunto o versión.');
 const clean={version:1,conjunto:D.conjunto,actual:v.actual,revisiones:{}};
 for(const [id,r] of Object.entries(v.revisiones)){
  if(!ids.includes(id)||!r||typeof r!=='object')throw Error('El archivo contiene una foto desconocida.');
  const n=blank();
  for(const [g,keys] of Object.entries(fields)){
   if(!r[g]||Object.keys(r[g]).length!==keys.length)throw Error('Faltan categorías o hay categorías desconocidas.');
   for(const k of keys){if(!allowed[g].includes(r[g][k]))throw Error('Una categoría tiene un valor inválido.');n[g][k]=r[g][k];}
  }
  if(typeof r.notas!=='string'||r.notas.length>3000||typeof r.completa!=='boolean'||typeof r.exposicion_previa!=='boolean')throw Error('Notas o estado de revisión inválidos.');
  n.notas=r.notas;n.completa=r.completa;n.exposicion_previa=r.exposicion_previa;
  for(const k of ['revelada_en','guia_en','terminada_en','actualizada_en']){if(r[k]!==null&&(typeof r[k]!=='string'||!Number.isFinite(Date.parse(r[k]))))throw Error('Fecha inválida.');n[k]=r[k];}
  if(!Array.isArray(r.historial)||r.historial.length>10000||r.historial.some(x=>!x||typeof x.accion!=='string'||x.accion.length>100||typeof x.fecha!=='string'||!Number.isFinite(Date.parse(x.fecha))))throw Error('Historial inválido.');
  n.historial=r.historial.map(x=>({accion:x.accion,fecha:x.fecha}));
  if(n.completa&&(!full(n)||!n.terminada_en))throw Error('Una revisión terminada tiene categorías sin responder.');
  clean.revisiones[id]=n;
 }
 return clean;
}
try{const saved=localStorage.getItem(KEY);if(saved)state=validate(JSON.parse(saved));}catch(e){storageBroken=true;message('No se pudo recuperar el guardado. No se reemplazará: exportá esta sesión y conservá cualquier copia anterior. '+e.message,true);}
function save(){
 if(externalConflict){message('Otra pestaña cambió la revisión. Exportá esta sesión antes de recargar; no se reemplazará el guardado de la otra pestaña.',true);return false;}
 if(storageBroken){message('Guardado automático no disponible. Exportá tu revisión antes de cerrar.',true);return false;}
 try{localStorage.setItem(KEY,JSON.stringify(state));message('Guardado en este navegador a las '+new Date().toLocaleTimeString('es-AR')+'.');return true;}catch(e){message('No se pudo guardar en el navegador. Exportá tu revisión antes de cerrar.',true);return false;}
}
window.addEventListener('storage',e=>{if(e.key===KEY&&e.newValue!==JSON.stringify(state)){externalConflict=true;save();}});
function revise(action,edit){const r=current();edit(r);r.actualizada_en=stamp();r.historial.push({fecha:r.actualizada_en,accion:action});state.revisiones[state.actual]=r;save();$('foto-estado').textContent=r.completa?'Revisión terminada.':'Borrador con cambios guardados. Terminá la revisión para incluirla en los puntajes.';progress();}
function expose(id,kind){const r=get(id),key=kind==='guia'?'guia_en':'revelada_en';if(!r[key]){r[key]=stamp();r.actualizada_en=r[key];r.historial.push({fecha:r[key],accion:kind==='guia'?'Consulta de guía':'Resultados revelados'});state.revisiones[id]=r;}}
function progress(){const n=Object.values(state.revisiones).filter(r=>r.completa).length;$('avance').textContent=n+' / '+ids.length;$('progreso').value=n;for(const o of $('foto-select').options)o.textContent=o.value+(get(o.value).completa?' · revisada':'');if(!$('puntajes').hidden)renderMetrics();}
function fieldHTML(g,k){const values=allowed[g].filter(Boolean),word={si:'Sí',no:'No',duda:'No puedo determinarlo',no_evaluable:'No evaluable'};return '<fieldset class="human-field"><legend>'+labels[k]+'</legend><div class="human-options">'+values.map(v=>'<label><input type="radio" name="'+g+'.'+k+'" value="'+v+'">'+word[v]+'</label>').join('')+'</div></fieldset>';}
for(const g of Object.keys(fields))$(g).innerHTML=fields[g].map(k=>fieldHTML(g,k)).join('');
$('foto-select').innerHTML=ids.map(id=>'<option value="'+id+'">'+id+'</option>').join('');
function exposureText(r){const tags=[];if(r.exposicion_previa)tags.push('Exposición previa declarada');if(r.guia_en)tags.push('Foto vista en la guía');if(r.revelada_en)tags.push('Resultados revelados');$('exposicion-aviso').textContent=tags.join('. ');}
function render(){
 const p=D.fotos.find(p=>p.foto===state.actual),r=current();shown=false;$('resultados-revelados').hidden=true;$('resultados-foto').replaceChildren();
 $('foto-select').value=state.actual;$('foto-humana').src=p.archivo;$('foto-humana').alt='Foto '+p.foto;$('foto-grande').href=p.archivo;$('foto-titulo').textContent=p.foto+' · tu evaluación';
 $('foto-estado').textContent=r.completa?'Revisión terminada. Si cambiás una respuesta, vuelve a borrador.':'Borrador. Respondé cada categoría para terminar; las dudas son válidas.';
 for(const [g,keys] of Object.entries(fields))for(const k of keys)for(const input of document.getElementsByName(g+'.'+k))input.checked=input.value===r[g][k];
 $('notas').value=r.notas;$('exposicion-previa').checked=r.exposicion_previa;exposureText(r);$('form-mensaje').textContent='';$('revelar').textContent='Revelar resultados de esta foto';
 $('anterior').disabled=ids.indexOf(state.actual)===0;$('siguiente').disabled=ids.indexOf(state.actual)===ids.length-1;progress();
}
function move(id){if(!ids.includes(id))return;state.actual=id;save();render();$('foto-grande').scrollIntoView({block:'start'});}
function nextPending(){const start=ids.indexOf(state.actual);for(let n=1;n<=ids.length;n++){const id=ids[(start+n)%ids.length];if(!get(id).completa){move(id);return;}}$('form-mensaje').textContent='Terminaste las '+ids.length+' fotos. Exportá una copia de tu revisión.';$('puntajes').hidden=false;renderMetrics();}
$('form-revision').addEventListener('change',e=>{if(e.target.type==='radio'){const [g,k]=e.target.name.split('.');revise('Cambio de categoría',r=>{r[g][k]=e.target.value;r.completa=false;});$('foto-estado').textContent='Borrador con cambios guardados. Terminá la revisión para incluirla en los puntajes.';}});
$('notas').addEventListener('input',()=>revise('Cambio de nota',r=>{r.notas=$('notas').value;r.completa=false;}));
$('exposicion-previa').addEventListener('change',()=>{revise('Exposición previa declarada',r=>{r.exposicion_previa=$('exposicion-previa').checked;r.completa=false;});exposureText(current());});
$('form-revision').addEventListener('submit',e=>{e.preventDefault();if(!full(current())){const missing=[];for(const [g,keys] of Object.entries(fields))for(const k of keys)if(!current()[g][k])missing.push(labels[k]);$('form-mensaje').textContent='Falta responder: '+missing.join(', ')+'. Podés marcar una duda.';return;}revise('Revisión terminada',r=>{r.completa=true;r.terminada_en=stamp();});nextPending();});
$('anterior').onclick=()=>move(ids[ids.indexOf(state.actual)-1]);$('siguiente').onclick=()=>move(ids[ids.indexOf(state.actual)+1]);$('pendiente').onclick=nextPending;$('foto-select').onchange=()=>move($('foto-select').value);
$('revelar').onclick=()=>{shown=!shown;$('resultados-revelados').hidden=!shown;$('revelar').textContent=shown?'Ocultar resultados':'Revelar resultados de esta foto';if(shown){expose(state.actual,'resultados');save();exposureText(current());const out=$('resultados-foto');out.replaceChildren();const row=$(state.actual);for(const node of row.querySelectorAll('.mode,.review')){const copy=node.cloneNode(true);copy.querySelectorAll('.comparison').forEach(n=>n.remove());out.append(copy);}out.scrollIntoView({block:'start'});}};
$('guia').addEventListener('toggle',()=>{if($('guia').open){for(const id of D.ejemplos||[])expose(id,'guia');save();exposureText(current());}});
$('ver-informe').onclick=()=>{for(const id of ids)expose(id,'resultados');save();$('human-app').hidden=true;$('legacy').hidden=false;window.scrollTo(0,0);};
$('volver-revision').onclick=()=>{$('legacy').hidden=true;$('human-app').hidden=false;render();window.scrollTo(0,0);};
function download(){const doc={...state,exportada_en:stamp(),criterios:'revision-humana-1',advertencia:'Referencia humana separada de la original. Material visible no equivale a servicio. Exposición registrada; no asumir revisión ciega.'};const a=document.createElement('a'),u=URL.createObjectURL(new Blob([JSON.stringify(doc,null,2)],{type:'application/json'}));a.href=u;a.download='ojo-urbano-revision-humana-'+new Date().toISOString().replace(/[:.]/g,'-')+'.json';document.body.append(a);a.click();a.remove();setTimeout(()=>URL.revokeObjectURL(u),30000);}
$('exportar').onclick=download;
$('importar').onclick=()=>$('archivo-revision').click();
const canonical=v=>Array.isArray(v)?v.map(canonical):v&&typeof v==='object'?Object.fromEntries(Object.keys(v).sort().map(k=>[k,canonical(v[k])])):v;
const equal=(a,b)=>JSON.stringify(canonical(a))===JSON.stringify(canonical(b));
function previewImport(parsed){const incoming=validate(parsed),conflicts=Object.keys(incoming.revisiones).filter(id=>state.revisiones[id]&&!equal(state.revisiones[id],incoming.revisiones[id]));pendingImport=incoming;const box=$('import-preview');box.hidden=false;box.replaceChildren();const p=document.createElement('p');p.textContent='Archivo válido: '+Object.keys(incoming.revisiones).length+' fotos guardadas. '+conflicts.length+' tienen diferencias con este navegador. Exportá tu revisión actual antes de reemplazar.';box.append(p);const button=(label,fn)=>{const b=document.createElement('button');b.textContent=label;b.onclick=fn;box.append(b);};button('Importar y conservar mis decisiones existentes',()=>mergeImport(false));if(conflicts.length)button('Reemplazar las decisiones en conflicto con las del archivo',()=>{download();mergeImport(true);});button('Cancelar',()=>{pendingImport=null;box.hidden=true;});}
function mergeImport(replace){for(const [id,r] of Object.entries(pendingImport.revisiones)){if(!state.revisiones[id]||replace){const old=state.revisiones[id];const n=structuredClone(r);if(old){n.revelada_en=old.revelada_en||n.revelada_en;n.guia_en=old.guia_en||n.guia_en;n.exposicion_previa=old.exposicion_previa||n.exposicion_previa;n.historial=[...old.historial,...n.historial,{fecha:stamp(),accion:'Importación con reemplazo'}];}state.revisiones[id]=n;}}state.actual=pendingImport.actual;pendingImport=null;$('import-preview').hidden=true;save();render();}
$('archivo-revision').onchange=async()=>{try{const f=$('archivo-revision').files[0];if(!f)return;if(f.size>5000000)throw Error('El archivo supera el tamaño admitido.');previewImport(JSON.parse(await f.text()));}catch(e){message('No se importó ningún dato: '+e.message,true);}finally{$('archivo-revision').value='';}};
const zero=()=>({tp:0,fp:0,fn:0,tn:0,pendientes_positivos:0,posibles:0,excluidos:0,confirmados_excluidos:0});
function score(keys,mode){const z=zero();for(const id of ids){const r=get(id);if(!r.completa)continue;const pred=D.modos[id][mode];for(const k of keys){const v=(services.includes(k)?r.servicios:r.contenedores)[k],yes=pred.confirmados.includes(k),possible=!yes&&pred.posibles.includes(k);if(v!=='si'&&v!=='no'){z.excluidos++;if(yes)z.confirmados_excluidos++;continue;}if(possible)z.posibles++;if(v==='si'){if(yes)z.tp++;else if(possible)z.pendientes_positivos++;else z.fn++;}else if(yes)z.fp++;else z.tn++;}}return z;}
const pct=(n,d)=>d?(100*n/d).toFixed(1)+'% ('+n+'/'+d+')':'Sin casos';
function table(groups){let h='<div class="human-breakdown"><table><thead><tr><th>Modo / categoría</th><th>Precisión</th><th>Detección</th><th>Confirmaciones incorrectas</th><th>Omisiones</th><th>Positivos pendientes</th><th>Posibles evaluables</th><th>Ejes excluidos</th></tr></thead><tbody>';for(const [title,keys] of groups)for(const mode of Object.keys(modes)){const s=score(keys,mode);h+='<tr><td>'+modes[mode]+' / '+title+'</td><td>'+pct(s.tp,s.tp+s.fp)+'</td><td>'+pct(s.tp,s.tp+s.fn+s.pendientes_positivos)+'</td><td>'+s.fp+'</td><td>'+s.fn+'</td><td>'+s.pendientes_positivos+'</td><td>'+s.posibles+'</td><td>'+s.excluidos+' ('+s.confirmados_excluidos+' confirmados)</td></tr>';}return h+'</tbody></table></div>';}
function renderMetrics(){const reviewed=ids.filter(id=>get(id).completa),n=reviewed.length;const exposed=reviewed.filter(id=>{const r=get(id);return r.exposicion_previa||(r.revelada_en&&r.revelada_en<=r.terminada_en)||(r.guia_en&&r.guia_en<=r.terminada_en);}).length;let h='<p><strong>'+n+' de '+ids.length+' fotos revisadas.</strong> '+exposed+' con exposición declarada o registrada antes de terminar la revisión. Las revisiones históricas ya vistas fuera de esta página deben declararse manualmente. No se asume que las demás sean ciegas.</p>';if(n!==ids.length)h+='<p class="human-warn">Muestra parcial elegida por el revisor. Todavía no representa el conjunto completo.</p>';h+=table([['Servicios',services],['Contenedores',containers]]);h+='<details><summary>Ver cada categoría</summary>'+table(Object.keys(labels).map(k=>[labels[k],[k]]))+'</details><h3>Costos conocidos y respuestas parciales</h3><ul>';for(const [mode,name] of Object.entries(modes)){const out=reviewed.map(id=>D.modos[id][mode]),costs=out.filter(r=>typeof r.costo==='number'),sum=costs.reduce((v,r)=>v+r.costo,0);h+='<li>'+name+': '+(costs.length?'USD '+(sum/costs.length).toFixed(6)+' promedio conocido en '+costs.length+' fotos':'sin fotos revisadas')+'. '+out.filter(r=>r.estado!=='completo').length+' respuestas parciales o sin verificación; '+out.filter(r=>r.consumo_completo===false).length+' con consumo incompleto.</li>';}h+='</ul><p class="human-muted">Los estados parciales se conservan. Los costos incompletos son mínimos conocidos. Dudas: no aumentan ni aciertos ni errores. Una categoría posible no se trata como confirmada.</p>';$('puntajes-contenido').innerHTML=h;}
$('ver-puntajes').onclick=()=>{$('puntajes').hidden=!$('puntajes').hidden;if(!$('puntajes').hidden){renderMetrics();$('puntajes').scrollIntoView({block:'start'});}};
render();if(!storageBroken)message('Revisión lista. Las decisiones se guardan automáticamente al responder.');
})();
